### PR TITLE
Feat/escrow contract

### PR DESCRIPTION
**Pull request by PabloLION**
This finish [DEV-48](https://abstrlabs.atlassian.net/browse/DEV-48)
~ 800 lines are auto generated / flowchart XML

## Feature

- algob-tester, with test : test algob capability and grammar
- Implement stake and unstake ART with PyTEAL (with test)
  - Mint aUSD by staking ART with a Gtxn consisted of 3 Txns
  - Burn minted aUSD to redeem ART, also with a Gtxn consisted of 3 Txns
  - SC throws error when burn>mint.

## Fix 
- v5 counter
- yarn command
- TealTester exporter

## Docs

- add note to README (moving to confluence)
- flowchart of the escrow smart contract
algorand-art-mint-SC

## Follow up

- Write "USE" in "README"

### CODE REVIEW

File: algo-builder-tester/assets/asa.yaml [hunk 1]
```diff
@@ -1,4 +1,33 @@
...
+$ART$:
+  total: 1e+12
+  decimals: 6
+  defaultFrozen: false
+  unitName: "ART"
+  url: "url"
+  metadataHash: "12345678901234567890123456789012"
+  note: "note"
+  noteb64: "noteb64"
+  # manager: "WWYNX3TKQYVEREVSW6QQP3SXSFOCE3SKUSEIVJ7YAGUPEACNI5UGI4DZCE"
```
Comment (bo-abstrlabs) [inline algo-builder-tester/assets/asa.yaml:10]: when these are not set, what are manager/reserve/freeze/clawback addr?

File: algo-builder-tester/test/v5escrow-test.ts [hunk 2]
```diff
@@ -0,0 +1,383 @@
...
+        fromAccount: alice.account,
+        toAccountAddr: billy.address,
+        payFlags: { totalFee: fee },
+      }
+      const receipt = runtime.executeTx(artPayTxParams) as typesRT.TxReceipt;
+      console.log('receipt : ', receipt); // DEV_LOG_TO_REMOVE
+      console.log('receipt.txn.snd : ', receipt.txn.snd); // DEV_LOG_TO_REMOVE
+      assert.equal(buf2Hex(receipt.txn.snd), u8a2Hex(parsing.addressToPk(alice.address)))
+      assert.equal(buf2Hex(receipt.txn.arcv!), u8a2Hex(parsing.addressToPk(billy.address)))
+      assert.equal(ab2Hex(receipt.txn.snd.buffer.slice(0, 32)), buf2Hex(receipt.txn.arcv!)) // Why??
```
Comment (bo-abstrlabs) [inline algo-builder-tester/test/v5escrow-test.ts:178]: another way to compare buffer/uint8array are https://www.chaijs.com/plugins/chai-bytes/, but buf2Hex etc. looks good too

File: algo-builder-tester/test/v5escrow-test.ts [hunk 3]
```diff
@@ -0,0 +1,383 @@
...
+        payFlags: { totalFee: fee },
+      }; // TODO:ref: store a basic transfer params
+      /* Check status before txn */
+      syncAccounts();
+      assert.equal(1000n, alice.assets.get(artID)!['amount']!); // 1k from dispense
+      assert.equal(999999998000n, admin.assets.get(artID)!['amount']!); // 2k dispensed
+      assert.equal(0n, alice.getLocalState(appID, ASSET_NAME)); // staked 0 $ART$ Unit
+      assert.equal(0n, alice.getLocalState(appID, STABLE_NAME)); // minted 0 aUSD Unit
+
+      const receipt = runtime.executeTx([mintCallParams, mintPayTxParams, mintCollectTxParams],);
```
Comment (bo-abstrlabs) [inline algo-builder-tester/test/v5escrow-test.ts:253]: Nice setup of grouped txn! what if user only do a mintCall, without mintpay and mintcollect?

File: algo-builder-tester/test/v5escrow-test.ts [hunk 4]
```diff
@@ -0,0 +1,383 @@
...
+      // console.log('receipt : ', receipt);
+
+      /* Check status after txn */
+      syncAccounts();
+      assert.equal(1000n, alice.assets.get(artID)!['amount']!);
+      assert.equal(999999998000n, admin.assets.get(artID)!['amount']!);
+      assert.equal(0n, alice.getLocalState(appID, ASSET_NAME)); // staked 0 $ART$ Unit
+      assert.equal(0n, alice.getLocalState(appID, STABLE_NAME)); // minted 0 aUSD Unit
+    });
+    it("burn>minted would fail", function () {
```
Comment (bo-abstrlabs) [inline algo-builder-tester/test/v5escrow-test.ts:313]: good test case!

File: artificial_algorand_contract/escrow/escrow_contract.py [hunk 5]
```diff
@@ -21,218 +27,246 @@
...
-        [Txn.on_completion() == OnComplete.CloseOut, handle_close_out],
-        [Txn.on_completion() == OnComplete.UpdateApplication, handle_update_app],
-        [Txn.on_completion() == OnComplete.DeleteApplication, handle_deleteapp],
-        [Txn.on_completion() == OnComplete.NoOp, handle_no_op],
+        [Gtxn[0].application_id() == Int(0), on_creation],
+        [Gtxn[0].on_completion() == OnComplete.NoOp, on_call],
+        [Gtxn[0].on_completion() == OnComplete.CloseOut, on_close_out],
+        [Gtxn[0].on_completion() == OnComplete.UpdateApplication, on_update_app],
+        [Gtxn[0].on_completion() == OnComplete.DeleteApplication, on_deleteapp],
+        [Gtxn[0].on_completion() == OnComplete.OptIn, on_opt_in],
```
Comment (bo-abstrlabs) [inline artificial_algorand_contract/escrow/escrow_contract.py:269]: Question: 
1) can Gtxn[1] Gtxn[2] ... also be CallApp type transaction
2) if 1) is true, can you specifiy another total 2048-byte arguments for each Gtxn
3) can you group more than 16 txns in TEAL version 6?

File: docs/algorand-art-mint-SC.drawio [hunk 6]
```diff
@@ -0,0 +1,258 @@
+<mxfile host="65bd71144e">
```
Comment (bo-abstrlabs) [inline docs/algorand-art-mint-SC.drawio:1]: Is this one same as you send me before. if not, could you please attach a screenshot in PR description?

**Review by bo-abstrlabs — APPROVED** [review]
Looks good to me

File: algo-builder-tester/test/v5escrow-test.ts (hunk 3 above)
Comment (PabloLION) [reply algo-builder-tester/test/v5escrow-test.ts:253]: Will be rejected by TEAL logic (length not 3). I didn't write this test. Will write one now.

File: algo-builder-tester/test/v5escrow-test.ts (hunk 2 above)
Comment (PabloLION) [reply algo-builder-tester/test/v5escrow-test.ts:178]: OK, I'll add a note for refactor.

File: algo-builder-tester/assets/asa.yaml (hunk 1 above)
Comment (PabloLION) [reply algo-builder-tester/assets/asa.yaml:10]: they are "undefined" I assume they are set to creator. I'll do a test now

File: artificial_algorand_contract/escrow/escrow_contract.py (hunk 5 above)
Comment (PabloLION) [reply artificial_algorand_contract/escrow/escrow_contract.py:269]: 1. Yes, they can . But the `Gtxn[0]` has to be the CallApp type (OnComplete.NoOp). I remembered seeing that the latter transaction cans see the new states of the first ones, but can't find the docs.
2. Yes.
3. No, you can't. see [Algorand parameter tables](https://developer.algorand.org/docs/get-details/parameter_tables/#others)

File: docs/algorand-art-mint-SC.drawio (hunk 6 above)
Comment (PabloLION) [reply docs/algorand-art-mint-SC.drawio:1]: No it's not. I'll try to add. I think there's a better way to insert than screenshot.

Comment (PabloLION) [reply docs/algorand-art-mint-SC.drawio]: it can't render the 20KB SVG. Uploaded a PNG at last

File: algo-builder-tester/assets/asa.yaml (hunk 1 above)
Comment (PabloLION) [reply algo-builder-tester/assets/asa.yaml:10]: > After an asset has been created only the manager, reserve, freeze and clawback accounts can be changed. All other parameters are locked for the life of the asset.
From [Algorand Standard Assets (ASAs)](https://developer.algorand.org/docs/get-details/asa/#modifying-an-asset)
So we cannot change the total number of coins. Maybe we publish more ART with less in circulation? I don't know how ALGO works.

Comment (PabloLION) [reply algo-builder-tester/assets/asa.yaml:10]: after test, I see they are set as "blank address" seems the creator can change all four accounts, if they are set to non-empty.

Comment (PabloLION) [reply algo-builder-tester/assets/asa.yaml:10]: test added: https://github.com/Artcoin-Network/artificial-algorand-contract/commit/adca6c30f1a471eb3896cfea7851b2a2c315ffcb

File: algo-builder-tester/test/v5escrow-test.ts (hunk 3 above)
Comment (PabloLION) [reply algo-builder-tester/test/v5escrow-test.ts:273]: I did it in algob-tester.ts tho. but it's not clearly shown here. adding a new one.

Comment (PabloLION) [reply algo-builder-tester/test/v5escrow-test.ts:274]: resolved https://github.com/Artcoin-Network/artificial-algorand-contract/commit/049055ba0f374c8f994bf038b1f137dc7ebeee3b

File: algo-builder-tester/test/v5escrow-test.ts (hunk 2 above)
Comment (PabloLION) [reply algo-builder-tester/test/v5escrow-test.ts]: resolved: https://github.com/Artcoin-Network/artificial-algorand-contract/pull/6/commits/97f2f5243bc3ae5d17d33ab46f66f9ad30e3db37

**Review by bo-abstrlabs — APPROVED** [review]
New test looks good!

**Pull request merged**